vector/read.go

```diff
@@ -14,570 +16,597 @@ import (
 	"github.com/brimdata/zed/zcode"
 )
 
-func Read(reader *vng.Reader, demandOut demand.Demand) (*Vector, error) {
-	context := zed.NewContext()
-	tags, err := readInt64s(reader.Root)
+func Read(object *vng.Object, demandOut demand.Demand) (*Vector, error) {
+	var buf []byte
+	tags, err := ReadInt64s(object.ReaderAt, &buf, object.Root)
 	if err != nil {
 		return nil, err
 	}
-	types := make([]zed.Type, len(reader.Readers))
-	values := make([]vector, len(reader.Readers))
-	for i, typedReader := range reader.Readers {
-		typeCopy, _ := context.DecodeTypeValue(zed.EncodeTypeValue(typedReader.Type))
-		typ := typeAfterDemand(context, typedReader.Reader, demandOut, typeCopy)
+	types := make([]zed.Type, len(object.Maps))
+	values := make([]vector, len(object.Maps))
+	for i, metadata := range object.Maps {
+		typeCopy := metadata.Type(object.Zctx)
+		typ := typeAfterDemand(object.Zctx, metadata, demandOut, typeCopy)
 		types[i] = typ
-		value, err := read(context, typedReader.Reader, demandOut)
+		value, err := read(object.Zctx, object.ReaderAt, &buf, metadata, demandOut)
```

Comment:
Nit: It looks as though the internal read functions (this one, ReadInt64s, readPrimitive, and readSegment at least) mostly share the same zed.Context, io.ReaderAt, and []byte parameters, so it might make things a little cleaner to stash them in a struct like
```go
type reader struct {
	zctx *zed.Context
	ra io.ReaderAt
	buf []byte
}
```
so you can hang the read functions on it as methods and then say
```go
		value, err := (&reader{object.Zctx, object.ReaderAt, nil}).read(metadata, demandOut)
```
here.